include/mbedtls/pk.h

```diff
@@ -484,6 +484,109 @@ int mbedtls_pk_can_do_ext(const mbedtls_pk_context *ctx, psa_algorithm_t alg,
                           psa_key_usage_t usage);
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
 
+#if defined(MBEDTLS_PSA_CRYPTO_CLIENT)
+/**
+ * \brief           Determine valid PSA attributes that can be used to
+ *                  import a key into PSA.
+ *
+ *                  The attributes determined by this function are suitable
+ *                  for calling mbedtls_pk_import_into_psa() to create
+ *                  a PSA key with the same key material.
+ *
+ *                  The typical flow of operations involving this function is
+ *                  ```
+ *                  psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+ *                  int ret = mbedtls_pk_guess_psa_attributes(pk, &attributes);
+ *                  if (ret != 0) ...; // error handling omitted
+ *                  // Tweak attributes if desired
+ *                  psa_key_id_t key_id = 0;
+ *                  ret = mbedtls_pk_import_into_psa(pk, &attributes, &key_id);
+ *                  if (ret != 0) ...; // error handling omitted
+ *                  ```
+ *
+ * \note            This function does not support RSA-alt contexts
+ *                  (set up with mbedtls_pk_setup_rsa_alt()).
+ *
+ * \param[in] pk    The PK context to use. It must have been set up.
+ *                  It can either contain a key pair or just a public key.
+ * \param[out] attributes
+ *                  On success, valid attributes to import the key into PSA.
+ *                  - The lifetime and key identifier are unchanged. If the
```

Comment:
As currently implemented, the lifetime and identifier are set to create a volatile key.

Initially, I had planned something different (and then I forgot to update this one sentence): the lifetime and key identifier are unchanged, and so you could call `set_key_{lifetime,identifier}` either before or after calling this function. This made the `MBEDTLS_PK_OPAQUE` key slightly less convenient: the code would have to save the lifetime/identifier fields and restore them after calling `psa_get_key_attributes` on the existing PSA key.

Which do you prefer @mpg?

Reply:
I don't think there's that much value in allowing `set_key_{lifetime,identifier}` either before or after compared to only after, so I'd go for the simplest to implement - that is, allowing it only after.

(I think it's even slightly easier on users, because that means all `psa_set_key_xxx()` functions that you call must occur after `guess_psa_attributes` and you don't have to remember which ones can go before and which ones must go after.)